### PR TITLE
[consensus] Update proposer metrics

### PR DESCRIPTION
**Pull request by arun-koshy**
## Description 

Add metric for the interval between propsals.

## Test plan 

[private-testnet](https://metrics.sui.io/goto/UhZ-qDkNR?orgId=1)

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

**Comment by arun-koshy** [comment]
All of the changes to the metrics can be seen reflected on the left side of the graphs and the right side is "main". Let me know what you think.

- We can see the rate at which we have to call back in to try new block because leaders were missing.
Screenshot 2024-10-02 at 10 55 09 AM
- We can see the the average wait time for a leader AFTER a quorum has been reached. With this we will have the quorum receive latency + leader wait time separated to show us which is taking most of the time.
Screenshot 2024-10-02 at 10 55 37 AM
- And this is what block proposal interval will look like.
Screenshot 2024-10-02 at 10 55 53 AM

File: consensus/core/src/core.rs [hunk 1]
```diff
@@ -275,11 +272,34 @@ impl Core {
     /// Adds/processed all the newly `accepted_blocks`. We basically try to move the threshold clock and add them to the
     /// pending ancestors list.
     fn add_accepted_blocks(&mut self, accepted_blocks: Vec<VerifiedBlock>) {
-        // Advance the threshold clock. If advanced to a new round then send a signal that a new quorum has been received.
-        if let Some(new_round) = self
-            .threshold_clock
-            .add_blocks(accepted_blocks.iter().map(|b| b.reference()).collect())
+        // Get max round of accepted blocks. This will be equal to the threshold
+        // clock round, either by advancing the threshold clock round by being
+        // greater than current clock round or by equaling the current clock round.
```
Comment (mwtian) [inline consensus/core/src/core.rs:277]: Is this the case? Blocks older than current threshold clock round can get accepted as well.

File: consensus/core/src/core.rs [hunk 2]
```diff
@@ -351,10 +371,22 @@ impl Core {
...
+            .hostname;
+
         // Create a new block either because we want to "forcefully" propose a block due to a leader timeout,
         // or because we are actually ready to produce the block (leader exists and min delay has passed).
         if !force {
             if !self.leaders_exist(quorum_round) {
+                self.context
+                    .metrics
+                    .node_metrics
+                    .block_proposal_leader_wait_count
```
Comment (mwtian) [inline consensus/core/src/core.rs:387]: I think we should use a separate metric for counting the number of times leader is not found. `block_proposal_leader_wait_count` is tied to `block_proposal_leader_wait_ms`, so when the average wait is ~250ms, we know the leader is missing.

File: consensus/core/src/core.rs (hunk 1 above)
Comment (arun-koshy) [reply consensus/core/src/core.rs:277]: I only added the case in the comment for greater and equal but blocks less than the clock round are essentially ignored by threshold clock

File: consensus/core/src/core.rs (hunk 2 above)
Comment (arun-koshy) [reply consensus/core/src/core.rs:387]: I think the confusion for me with these metrics is that it doesn't just include leader wait time, it includes the quorum receive wait time which can make this metric a little misleading. Separating them brings more clarity. Though I guess we could always subtract this metric from quorum receive latency.

Comment (mwtian) [reply consensus/core/src/core.rs:387]: Looking at `ThresholdClock::add_block()`, `quorum_ts` is updated when a quorum forms. Then the `block_proposal_leader_wait_ms` is calculated from `quorum_ts`. This does not include quorum receive latency. Also, leader timeout task is signaled when `quorum_ts` is updated, so `block_proposal_leader_wait_ms` should line up with actual leader wait.

File: consensus/core/src/core.rs [hunk 3]
```diff
@@ -477,6 +500,19 @@ impl Core {
...
+        let current_proposal_duration = Duration::from_millis(verified_block.timestamp_ms());
+        let previous_proposal_duration = Duration::from_millis(self.last_proposed_timestamp_ms());
+        self.context
+            .metrics
+            .node_metrics
+            .block_proposal_interval
+            .observe(
+                current_proposal_duration
+                    .checked_sub(previous_proposal_duration)
+                    .unwrap_or_else(|| Duration::from_millis(0))
```
Comment (mwtian) [inline consensus/core/src/core.rs:512]:
```suggestion
                    .saturating_sub(previous_proposal_duration)
```

File: consensus/core/src/metrics.rs [hunk 4]
```diff
@@ -232,6 +233,12 @@ impl NodeMetrics {
                 &["authority"],
                 registry,
             ).unwrap(),
+            block_proposal_interval: register_histogram_with_registry!(
```
Comment (mwtian) [inline consensus/core/src/metrics.rs:236]: This metric is useful.

File: consensus/core/src/threshold_clock.rs [hunk 5]
```diff
@@ -24,33 +26,59 @@ impl ThresholdClock {
...
+
+    fn add_block(&mut self, block: BlockRef, proposal_leaders_exist: bool) {
         match block.round.cmp(&self.round) {
             // Blocks with round less then what we currently build are irrelevant here
             Ordering::Less => {}
             // If we processed block for round r, we also have stored 2f+1 blocks from r-1
             Ordering::Greater => {
                 self.aggregator.clear();
                 self.aggregator.add(block.author, &self.context.committee);
+                if proposal_leaders_exist {
```
Comment (mwtian) [inline consensus/core/src/threshold_clock.rs:68]: I feel if we want to monitor the time between round N having quorum and round N leader exists, doing it in core is more natural, and it is almost equivalent to how `block_proposal_leader_wait_ms` is computed.

**Review by mwtian — COMMENTED** [review]
My suggestion is to keep the `block_proposal_interval` metric and postpone the refactor to `block_proposal_leader_wait_ms` and `block_proposal_leader_wait_count`. We can chat offline on the goal of the refactor and the best way to achieve that.

File: consensus/core/src/core.rs (hunk 2 above)
Comment (arun-koshy) [reply consensus/core/src/core.rs:387]: What I was trying to measure is the specific moment when the leader block is received versus when the round can advance due to quorum received. The wait time is anytime after the quorum is received as that is the minimum needed to advance the round. The leader block can be received and make up the quorum needed to advance the clock. So to accurately get the wait time for a leader block we should get the `min(leader_received_ts - quorum_ts, 0)`. 

What we currently have in theory should lead to the same results but after we add the block we go through the process of `try_commit` which then adds to the wait time calculation for the leader. I think this is why we see the difference in the graphs. (pasting here again)

[left is the new metric and right is main]
image

Comment (mwtian) [reply consensus/core/src/core.rs:387]: Is the average is calculated with block_proposal_leader_wait_ms / block_proposal_leader_wait_count, can you compare the `block_proposal_leader_wait_count` rate as well? Originally the intention is to have the count similar to # of blocks proposed. With the new logic, I think it is double counting: if a block proposal has to wait for 250ms, and `try_new_block()` gets called after waiting for 50ms, 100ms, 150ms and 200ms, then the average leader wait seems to be skewed higher.

Comment (arun-koshy) [reply consensus/core/src/core.rs:387]: Yeah you are correct using leader count was incorrect in the new metric but comparing it looks like the wait is indeed lower still

https://metrics.sui.io/goto/mZanqCiHg?orgId=1

**Review by mwtian — APPROVED** [review]

**Pull request merged**